### ISSUE TITLE
Add password validation during user creation

### ISSUE THREAD

**Issue by conorsch** [freedomofpress/securedrop#980]
Currently the application permits empty strings for passwords during account creation. This affects account creation via both `manage.py` in the CLI and the web interface. At a minimum the app should check for empty strings and prompt again. Better yet, the app should enforce a minimum password strength via regex.

We also need tests to confirm this functionality (rejecting weak passwords) once it's added.

**Comment by abbynormative** [comment]
I added a quick fix for this but you may want to use Cracklib instead or as well because they have a dictionary checker. https://pypi.python.org/pypi/cracklib

**Comment by garrettr** [comment]
This needs some more unpacking for us to find an effective solution. I am wary of "password strength" metrics in general because they usually over-estimate password strength and often impose barriers to usability.

For example:

1. Checking for a minimum length is a good idea, but is not enough by itself. A password may be long but easily cracked by a dictionary attack (e.g. `qwerytasdfgzxcvb`).
2. Checking for a diversity of character types (upper/lower/number/symbol) helps in theory, but:
    1. It imposes barriers to memorability that often lead humans to choose passwords that look good but are actually vulnerable to dictionary attacks, or choose passwords that are way too hard to remember (https://xkcd.com/936/).
    2. Disallows passwords that may not have a diversity of character types, but are still highly secure (e.g. Diceware passphrases, or long random passphrases that are only alpha characters to aid in typing, especially on touchscreens).

Our documentation tells journalists and admins to create this password using KeePassX, so it will be long, random, and contain mixed character types (default is 25 characters of lower/upper/number/symbol). It seems reasonable to make this *recommendation* a **requirement** in the password validation code. Doing so creates a minor inconvenience for application developers (who have to log in to the Document Interface frequently while testing), but after some debate with @dolanjs I think this is worth it for the higher level of security assurance.

The best way to implement this is probably to validate for passwords that match the default settings in KeePassX on Tails (which is what journalists and admins use for their workstations). The tricky thing in implementing this will be dealing with the fact that a randomly chosen password *might not* contain characters from every group. KeePassX has a checkbox to "Ensure that password contains characters from every group", but it is not on by default.

![screen shot 2015-04-10 at 5 46 14 pm](https://cloud.githubusercontent.com/assets/407302/7099532/8c941ba4-dfa9-11e4-8140-795a6d48f33d.png)

**Comment by abbynormative** [comment]
Good to know. I'll think about it and try to have something better next week. Thanks Garrett!

**Comment by abbynormative** [comment]
This is the solution I'm thinking about, however I'm concerned about only checking that it contains at least one of the recommended character groups:

while True:
    print """Please enter a password.

We recommend using KeePassX to generate a random password of at least 25
characters containing a combination of lower/upper case, numbers, and symbols.
"""
    password = getpass("Password: ")
    password_again = getpass("Confirm Password: ")
    if password == password_again:
        print "Your password is " + str(len(password)) + " characters long."
        if len(password) >= 25:
            if re.search('\d+', password) or re.search('[a-z]', password) or re.search('[A-Z]', password) or re.search('[^\w\s]', password):
                print """Your password also contains the recommended mix of characters and passes our 
                strength test."""
                break
            else:
                print "Your password is not strong enough. Please try again."
        else:
            print "Your password is less than 25 characters. Please try again."
    else:
        print "Passwords didn't match!"

**Comment by garrettr** [comment]
@abbynormative That looks pretty good. Regarding checking for the multiple character groups, the edge case that I'm worried about is:

1. User follows best practice and generates a random KeePassX password; however the password doesn't have characters from all of the groups (because "Ensure that password contains characters from every group" is not checked by default).
2. They enter it into the prompt (assuming we change the validation requirements to so 4/4 or 3/4 characters groups must be present).
3. It fails, and we've confused them and also punished them for following best practices :cry: 

It seems like checking for 3/4 groups might be an ok compromise, but you would still encounter this edge case with some probability.

**Comment by garrettr** [comment]
Another option here is to automatically generate a passphrase for them, and instruct them to copy it into their password database. That way there can be no confusion :).

If we do this, I recommend optimizing the password for:

1. Strength
2. Ease of typing
3. Ease of spotting errors

While they will typically be copy/pasting the password from KeePassX (and so won't need to type it), I still think it's best to optimize passwords for humans in case there are issues with the copy/paste and the user has no choice but to manually type in the password.

My preferred way to do this is to base32-encode some entropy from a strong CSPRNG (e.g. `/dev/urandom`). This ensures a very strong passphrase (assuming the CSPRNG is good) but also makes it easy to type (base32 is optimized for human readability and manual entry).

```py
from base64 import b32encode
import os

def generate_friendly_password():
    # 16 bytes of min-entropy == 128 bit security against brute force.
    # Passwords are 26 characters long
    return b32encode(os.urandom(16)).lower().strip('=')
```

128 bits of min-entropy is excessive, but since they are going to be copy/pasting this from a password DB most of the time, I think it is acceptable.

**Comment by dolanjs** [comment]
@garrettr I think using diceware pass-phrases meets your concerns about strength, ease of typing, ease of spotting errors concerns. I do think this will take more discussion/planning since we could improve the current processes a lot.

Currently we have two use cases to cover
## First journalist user
* The first journalist user (also the default admin user) is created on the app servers command line by the server admin.
* The server admin creates the initial document interface user account, initial password and copies 2fa value.
* The server admin transfers the document interface user's 2fa secret value, password to the journalist via sneakernet or gpg email.
* The journalist logins to the document interface and changes their password and 2fa value.

## Additional journalist users
* The journalist admin logins to document interface and clicks the admin link in the top right corner
* The journalist admin creates a new journalist user with 2fa creds and passphrase and transfers those creds via sneakernet or gpg email to the new journalist user.
* The new journalist user logins to the document interface and changes creds.

## Current weak areas.
* Journalist users who are not Document Interface admins can not reset their own 2fa code or password.
* Users will create simple passwords that get around password checks one way or the other.
* The person who initially creates the user has access to the 2fa secret and initial password of the users they create.
* We are trusting that the admin securely transfers the 2fa secret and passphrase to the journalist user.
* We are counting on the journalist actually resetting their initial password and 2fa value to ensure that the admin user does not have access to their creds.

**Comment by JMPorup** [comment]
@garrettr I wonder if that code could be abstracted to lay the groundwork for future language support.

What do you think of this idea?

```python
from base64 import b32encode
import os

@generate_local_alphabet
def generate_friendly_password():
    # 16 bytes of min-entropy == 128 bit security against brute force.
    # Passwords are 26 characters long
    return os.urandom(16)

def generate_local_alphabet(f, locale="en_US")
    #locale a global variable? or passed in?
    if locale == "en_US"
        return b32encode(f).lower().strip('=')
    elif locale == "de_DE"
        return something else
    elif locale == "cn_CN"
    #how to deal with Asian language support?
        &etc...
```
I've probably made a hash of this, but you get the idea. :grimacing:

**Comment by fowlslegs** [comment]
It believe it would be simpler to add text to `manage.py` and the Document Interface that instructed the user to generate a password using KeePassX, since we already have documentation instructing journalists on use of KeePassX, and teach it in our in-person trainings. Password validation could require a minimum length, and have a loose validation function, that will tell us that the password was at least _probably_ generated by KeePassX, while avoiding throwing errors with any significant frequency. I think it would be possible to write such a validator that would only throw an error for 1 in 1,000 legitimate KeePassX passwords, while still providing some checks. There is no perfect password strength validator, and we will never be able to strongly protect the security of Journalist Workstations, when journalists ignore explicit instructions. They are probably cutting corners for convenience in many other ways if they won't take the 30 seconds extra time it takes to use KeePassX.

**Comment by heartsucker** [comment]
@fowlslegs @conorsch Just checking in here before I try again to implement this. The new feature will:

1) Auto generate a password on user creation
2) Assign a new password on first login

Both 1 and 2 would use the current diceware list for passwords (n = 7). The DB would be changed to have a column for jounalist like `require_pw_reset` that triggers redirects to a password reset. This would also be a nice reason to implement #1419.

**Comment by fowlslegs** [comment]
> Assign a new password on first login

Not sure what you mean here?

> The DB would be changed to have a column for jounalist like require_pw_reset that triggers redirects to a password reset.

Also confused what you mean here.

...

To clarify and expand on what I was suggesting:

1. In both the `manage` module and the GUI interface user creation--both w/ and w/o admin privileges--would no longer prompt for a password. Instead, a 7-word Diceware passphrase will be generated and displayed to the user (I'm going to make an issue soon explaining why I think we should get eliminate the ability for sources to choose `n`, but that's a different story :smile:).
2. In the account settings view of the journalist app we'll need to change the mechanism for changing an user's password. Much like the button which changes the `source.journalist_id`, a "change password" POST/ button can exist which performs a similar function. To prevent accidental account lockouts, confirmation of the new password via typing/ copying it into a form field should be necessary.

Two more things to note:

1. Existing users in upgraded instances should _not_ have to update their passwords.
2. Appropriate messages reminding journalists/admins to save their new password in the KeePassX database on their journalist/admin workstations should be added to `manage.py`/`journalist.py`.

You should talk to @redshiftzero more about https://github.com/freedomofpress/securedrop/issues/1419, but I don't think DB changes should be necessary for this one.

**Comment by fowlslegs** [comment]
Sticking this in 0.4 since we have a pending PR, but this might be too big of a change for our scope of that release. Should be discussed at our next meeting.

**Comment by heartsucker** [comment]
@fowlslegs All good. Rebasing this one isn't too bad. :)

**Comment by redshiftzero** [comment]
Closed by #1509.
